src/Backups/BackupIO.h

```diff
@@ -38,7 +42,17 @@ class IBackupWriter /// BackupWriterFile, BackupWriterDisk
     virtual DataSourceDescription getDataSourceDescription() const = 0;
     virtual void copyDataToFile(const CreateReadBufferFunction & create_read_buffer, UInt64 offset, UInt64 size, const String & dest_file_name);
     virtual bool supportNativeCopy(DataSourceDescription /* data_source_description */) const { return false; }
+
+    /// Copy file using native copy (optimized for S3 to use CopyObject)
+    ///
+    /// NOTE: It still may fall back to copyDataToFile() if native copy is not possible:
+    /// - different buckets
+    /// - throttling had been requested
     virtual void copyFileNative(DiskPtr src_disk, const String & src_file_name, UInt64 src_offset, UInt64 src_size, const String & dest_file_name);
+
+protected:
+    const ReadSettings read_settings;
```

Comment:
First I thought you're adding a throttler for writing, not for reading. That's why I suggested moving `read_settings` here. But since now I know you're adding a throttler for reading that suggestion doesn't look relevant anymore. I mean it's the `IBackupWriter` interface, it's not `IBackupReader`, and it's a bit strange to have `read_settings` here. Yes, this class has the `copyFileNative()` function, but that doesn't seem to be an integral part of the `IBackupWriter` interface. So it seems it's better to move `ReadSettings` back to the arguments of `copyDataToFile` and `copyFileNative`. I don't think we need `has_throttling` there because it can be extracted from `ReadSettings`.

Reply:
Yes, this is throttling for reads (write side will be throttled automatically by reads anyway).

I'm OK with changing this interface as you suggested, although I would prefer to keep it as a member of `IBackupWriter`, yes, it is `ReadSettings`, but it is common, all writers needs it, because they are not only writes, they actually read->write, and they need it for two things: check throttling and pass `read_settings` in case of native copy falls back to regular copy.

Also initially `BackupIO_S3.cpp` has `ReadSettings` as a member - https://github.com/ClickHouse/ClickHouse/blob/fa5b2bd4a5b02336bca8837c473a7124f8ecedf2/src/Backups/BackupIO_S3.h#L79

So just to double check, are you sure that you want this to be moved to the arguments of functions?

Reply:
I see `ReadSettings` is rather a context-specific thing than a read-operation specific thing. I mean it's the same for the Context, it is not generated for each read operation. So it's like if we kept a context in BackupWriterS3.